File: parallel.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <omp.h>

#ifndef MAX_FILAS
#define MAX_FILAS 1100
#endif
#ifndef MAX_COLUMNAS
#define MAX_COLUMNAS 1100
#endif

static char mapa[MAX_FILAS][MAX_COLUMNAS];
static int ancho_fila[MAX_FILAS];

int main() {
    int filas = 0;
    int i, j;
    int contador_total = 0;
    int num_hilos = 0;
    double inicio, fin;

    FILE *archivo = fopen("mapa_grande.txt", "r");
    if (!archivo) { printf("Error: no se pudo abrir mapa_grande.txt\n"); return 1; }

    char linea[MAX_COLUMNAS + 2];
    while (filas < MAX_FILAS && fgets(linea, sizeof(linea), archivo)) {
        linea[strcspn(linea, "\n")] = '\0';
        int columnas_leidas = (int)strlen(linea);
        if (columnas_leidas > MAX_COLUMNAS) columnas_leidas = MAX_COLUMNAS; // <-- tope

        for (j = 0; j < columnas_leidas; j++) {
            mapa[filas][j] = linea[j];
        }
        for (j = columnas_leidas; j < MAX_COLUMNAS; j++) {
            mapa[filas][j] = 0;
        }

        ancho_fila[filas] = columnas_leidas; 
        filas++;
    }
    fclose(archivo);

    inicio = omp_get_wtime();

    #pragma omp parallel
    {
        int tid = omp_get_thread_num();
        int contador_local = 0;

        #pragma omp single
        num_hilos = omp_get_num_threads();

        #pragma omp for
        for (i = 0; i < filas; i++) {
            int cols = ancho_fila[i]; // <-- usa el ancho correcto de esa fila
            for (j = 0; j < cols; j++) {
                if (mapa[i][j] == 'X') contador_local++;
            }
        }

        #pragma omp atomic
        contador_total += contador_local;

        #pragma omp critical
        printf("Hilo %d encontró %d tesoros en su sección.\n", tid, contador_local);
    }

    fin = omp_get_wtime();

    printf("\nParalelo: Total de tesoros = %d\n", contador_total);
    printf("Cores utilizados: %d\n", num_hilos);
    printf("Tiempo de ejecucion: %f segundos\n", fin - inicio);
    return 0;
}

File: parallel.py

from concurrent.futures import ProcessPoolExecutor, as_completed
from math import ceil
import os
import time
import sys

FILENAME = "mapa_grande.txt"
MAX_FILAS = 1100
MAX_COLUMNAS = 1100

def leer_mapa(path: str):
    try:
        with open(path, "r", encoding="utf-8") as f:
            lineas = []
            for idx, raw in enumerate(f):
                if idx >= MAX_FILAS:
                    break
                linea = raw.rstrip("\n")
                # recortar si supera MAX_COLUMNAS
                if len(linea) > MAX_COLUMNAS:
                    linea = linea[:MAX_COLUMNAS]
                lineas.append(linea)
            return lineas
    except FileNotFoundError:
        print("Error: no se pudo abrir", path)
        sys.exit(1)

def contar_en_bloque(bloque, worker_id):
    # Cuenta 'X' en un bloque de filas. Retorna (id, conteo_local)
    contador_local = 0
    for fila in bloque:
        contador_local += fila.count('X')
    msg = f"Hilo {worker_id} encontró {contador_local} tesoros en su sección."
    return worker_id, contador_local, msg

def dividir_en_bloques(filas, num_workers):
    n = len(filas)
    if n == 0:
        return []
    tam = ceil(n / num_workers)
    bloques = []
    start = 0
    worker_id = 0
    while start < n:
        bloques.append((filas[start:start+tam], worker_id))
        start += tam
        worker_id += 1
    return bloques

def main():
    mapa = leer_mapa(FILENAME)
    filas = len(mapa)
    columnas = max((len(s) for s in mapa), default=0)

    # Número de procesos (como "num_hilos" en OpenMP).
    num_workers = os.cpu_count() or 1

    inicio = time.perf_counter()

    # Partimos el trabajo por bloques de filas
    bloques = dividir_en_bloques(mapa, num_workers)

    contador_total = 0
    mensajes = []
    # Ejecutamos en paralelo
    with ProcessPoolExecutor(max_workers=num_workers) as ex:
        futuros = [ex.submit(contar_en_bloque, bloque, wid) for (bloque, wid) in bloques]
        for fut in as_completed(futuros):
            wid, local_count, msg = fut.result()
            contador_total += local_count
            mensajes.append((wid, msg))

    fin = time.perf_counter()

    for _, msg in sorted(mensajes, key=lambda x: x[0]):
        print(msg)

    print(f"\nParalelo: Total de tesoros = {contador_total}")
    print(f"Cores utilizados: {num_workers}")
    print(f"Tiempo de ejecucion: {fin - inicio:.6f} segundos")
    print(f"Filas leídas: {filas}, Columnas (máx por fila): {columnas}")

if __name__ == "__main__":
    main()
